### DISCUSSION TITLE
Applying rule UP032 in favour of UP030/31

### DISCUSSION THREAD

**Discussion by ihavran** [astral-sh/ruff#7810] [Q&A] [answered]
Hey,

as I was recently upgrading codebase, I became aware of unusual behavior in certain rules in **pyupgrade**. I didn't find a thread about similiar topic, so I decided to open one here.

When I applied rule **UP030/31**, I expected rule **UP032** to be applied in next call I made. But this wasn't the case at all times - actually it was pretty inconsistent when it was applied (I'm not aware of any specific rule).

There is not additional information about applying these rules together, but my assumption was, that rules **UP030/31** breaks rule **UP032** and therefore change should be applied.

Is there some explanation for this behavior?

Thanks in advance

**Comment by charliermarsh** [comment]
Before I respond in more detail, are you able to provide an example of a string that was migrated via 30/31, but not 32?

**Reply by ihavran** [reply]
Hey @charliermarsh , I'll provide some examples. Just to be on same page, usual routine for upgrade was like this:

- ruff --select "UPXYZ" --fix where XYZ were all rules that got found by initial scan (following order, so 32 got applied after 30/31 at all times)
- ruff --select "UP" --fix to be sure everything got applied

This is a example where **UP031 got applied, but UP032 did not**.

**Before:**
raise Error("ID (%s=%s) excluded!" % (object.get("id_key"), object.get("id_value"))

**After:**
raise Error("ID ({}={}) excluded!".format(object.get("id_key"), object.get("id_value"))

Another example, **UP030 got applied, UP032 didn't**.

**Before:**
"_accepts_marketing_{0}".format(self.domain_safe)

**After:**
"_accepts_marketing_{}".format(self.domain_safe)

Last example, where **UP030 got applied and also UP032 shortly after**

**Before:**
"{0}_{1}".format(prefix, self.market_domain)

**After:**
f"{prefix}_{self.market_domain}"

I hope this helps

**Reply by charliermarsh** [reply]
Thank you!

In the `accepts_marketing` example, running `ruff check /path/to/file.py --select UP --fix`, I do go from:

```python
"accepts_marketing{0}".format(self.domain_safe)
```

To:

```python
f"accepts_marketing{self.domain_safe}"
```

That one seems to be "working as intended" for me -- can you confirm?

In the third example, that's also "working as intended", right? Since it _did_ get converted to an f-string?

The first example I agree would ideally convert to an f-string. It's tricky due to the nested quotes.

**Reply by ihavran** [reply]
I confirm, I had to change some parts to be able for publish, so maybe that's the reason...

Yes, the third example was correct usage - sorry, I could have explicitly said that.

I looked for some more examples - there is this one, which doesn't get detected at all:

`return "<Customer: {}, {}, {}>".format(self.interal_id, self.external_id, self.property)`

Is there some issue with detecting when specific symbols are used (like <> here or nested quotes you mentioned)? I'll be allright with low-level explanation as well - just want to inform other developers about the limitation (and possible cause)

**Reply by charliermarsh** [reply]
Hmm, that one works for me too -- notice that it's flagged as a violation here: https://play.ruff.rs/0168e31d-fc7a-4049-b0ab-2c335f918ed5.

One case we _don't_ fix, if it's relevant: if you have any comments within the `()` in `format(...)`.

**Reply by ihavran** [reply]
Sorry, seems like when you test the string it's working - so I'm copying whole class (removed some parts)

https://play.ruff.rs/2522653d-a2fd-40a3-a9ff-664d58d159b8

Seems like it doesn't get flagged in this case

**Reply by charliermarsh** [reply]
Thank you! I think that's just a bug in the rule. Can fix.

**Reply by ihavran** [reply]
thanks @charliermarsh !

**Reply by charliermarsh** [reply]
The reason this isn't being applied (which isn't clear) is that it would cause the string to exceed the configured line width. (If you increase the line width to, like, 120, the error is restored.) In this example, the string _already_ exceeds the line width, so we should flag it anyway...

**Reply by charliermarsh** [reply, marked as answer]
Okay, that specific issue has been resolved on `main`. Hopefully that helps!

**Reply by ihavran** [reply]
Thanks again for explanation and quick resolve :pray:
